addon/PreviewProvider.js

```diff
@@ -359,15 +359,34 @@ PreviewProvider.prototype = {
   },
 
   /**
-   * Do some pre-processing on the links before inserting them into the metadata
+   * Do some pre-processing on the link before inserting it into the metadata
    * DB and adding them to the metadata cache
    */
-  processAndInsertMetadata(links) {
-    const processedLinks = this._processLinks(links);
-    this.insertMetadata(processedLinks);
+  processAndInsertMetadata(link, metadataSource) {
+    const processedLink = this._processLinks(link);
+    this.insertMetadata(processedLink, metadataSource);
   },
 
   /**
+   * Check if a single link exists in the metadata DB
+   */
+  asyncDoesSingleLinkExist: Task.async(function*(url) {
+    let link = this._processLinks([{url}])[0];
+    if (!link) {
+      return false;
+    }
+    // Hit the cache first
+    const key = link.cache_key;
+    const metadataFromCache = MetadataCache.cache.get(key);
+
```

Comment:
Short circuit with `return true` on success should flow better here while unindenting the cache miss case.